### PR TITLE
Enhancements around Zulip APIs

### PR DESCRIPTION
**Pull request by LeSeulArtichaut**
This PR proposes some enhancements for sending and receiving messages, and adds a new `link` commands which sends back a link to the current channel (topic or private message).

- The Zulip `MessageAPIRequest` is now more statically typed and uses an enum to determine the type of recipient, i.e. `"private"` or `"stream"`.
- The `Message` data structure now parses more information, such as the `stream_id` and `topic` of the message. I didn't test this but only tried to construct a valid data structure from the [Zulip API documentation examples](https://www.zulipchat.com/api/get-events#example-response). I think a second look is necessary.

The `link` command is meant as a helper for e.g. mobile clients which have no easy way to determine the link to a topic. Its implementation could also be reused in the context of other commands.

cc @Mark-Simulacrum

### CODE REVIEW

File: src/zulip.rs [hunk 1]
```diff
@@ -303,48 +305,94 @@ struct MembersApiResponse {
...
-        let always_safe = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_.-~";
-        let mut encoded_topic = Vec::new();
-        for ch in self.topic.expect("topic").bytes() {
-            if !(always_safe.contains(ch as char) || ch == b'.') {
-                write!(encoded_topic, "%{:02X}", ch).unwrap();
-            } else {
-                encoded_topic.push(ch);
-            }
+impl<'a> MessageApiRequest<'a> {
+    pub fn private(user_id: u64, user_email: &'a str, content: &'a str) -> Self {
```
Comment (Mark-Simulacrum) [inline src/zulip.rs:368]: I am not a fan of this as it has two &str fields without much protection for getting it right, and the u64 is also not obviously just one thing. I would prefer that we construct the ApiRequest directly at the call sites for both of these constructor functions.

File: src/zulip.rs [hunk 2]
```diff
@@ -520,3 +568,21 @@ async fn move_notification(
         .unwrap()),
     }
 }
+
+fn message_channel_link(msg: &Message) -> String {
```
Comment (Mark-Simulacrum) [inline src/zulip.rs:572]: Can we at least split this addition to a separate commit? I'm also not clear on the use case--Zulip makes it pretty easy to get links, no?

Comment (LeSeulArtichaut) [reply src/zulip.rs:572]: Here were my thoughts on this (from this PR's comment):

> The `link` command is meant as a helper for e.g. mobile clients which have no easy way to determine the link to a topic. Its implementation could also be reused in the context of other commands.

Comment (LeSeulArtichaut) [reply src/zulip.rs]: Removed it and will add it back later
